plugins/window-rules/window-rules.cpp

```diff
@@ -0,0 +1,234 @@
+    output->disconnect_signal("view-fullscreen", &_fullscreened);
+}
+
+void wayfire_window_rules_t::apply(const std::string & signal,
+    wf::signal_data_t *data)
+{
+    if (data == nullptr)
+    {
+        return;
+    }
+
+    auto view = get_signaled_view(data);
+    if (view == nullptr)
+    {
+        LOGE("View is null.");
+
+        return;
+    }
+
+    if ((signal == "maximized") && (view->tiled_edges != wf::TILED_EDGES_ALL))
+    {
+        return;
+    }
+
+    if ((signal == "unmaximized") && (view->tiled_edges == wf::TILED_EDGES_ALL))
+    {
+        return;
+    }
+
+    for (const auto & rule : _rules)
+    {
+        _access_interface.set_view(view);
+        _action_interface.set_view(view);
+        auto error = rule->apply(signal, _access_interface, _action_interface);
+        if (error)
+        {
+            LOGE("Window-rules: Error while executing rule on ", signal, " signal.");
+        }
+    }
+
+    auto bounds = _lambda_registrations->rules();
+    auto begin  = std::get<0>(bounds);
+    auto end    = std::get<1>(bounds);
+    while (begin != end)
+    {
+        auto registration = std::get<1>(*begin);
+        bool error = false;
+
+        // Assume we will use the view access interface.
+        _access_interface.set_view(view);
+        wf::access_interface_t & access_iface = _access_interface;
+
+        // If a custom access interface is set in the regoistration, use this one.
+        if (registration->access_interface != nullptr)
+        {
+            access_iface = *registration->access_interface;
+        }
+
+        // TODO: Remove debug/test code.
+        LOGW("Apply rule to view: ", view->get_app_id(), ": ");
+
+        // Load if lambda wrapper.
+        if (registration->if_lambda != nullptr)
+        {
+            registration->rule_instance->setIfLambda(
+                [registration, signal, view] () -> bool
+            {
+                if (registration->if_lambda != nullptr)
+                {
+                    return registration->if_lambda(signal, view);
+                }
+
+                return false;
+            });
+        }
+
+        // Load else lambda wrapper.
+        if (registration->else_lambda)
+        {
+            registration->rule_instance->setElseLambda(
+                [registration, signal, view] () -> bool
+            {
+                if (registration->else_lambda != nullptr)
+                {
+                    return registration->else_lambda(signal, view);
+                }
+
+                return false;
+            });
+        }
+
+        // Run the lambda rule.
+        error = registration->rule_instance->apply(signal, _access_interface);
+
+        // Unload wrappers.
+        registration->rule_instance->setIfLambda(nullptr);
+        registration->rule_instance->setElseLambda(nullptr);
+
+        if (error)
+        {
+            LOGE("Window-rules: Error while executing rule on ", signal, " signal.");
```

Comment:
Error can be set true in the following conditions:
* The lambda returns true to signal an error during the execution of whatever code is in the lambda.
* The access interface returns an error because it is looking for a property that does not exist (in the case a rule mentions a non-existing variable, but in a semantically correct way).

Reply:
Alright, then it may make sense to also log the whole rule text, if possible.

Reply:
Done